### ISSUE TITLE
Port old prusti-filter checks

### ISSUE THREAD

**Issue by fpoli** [viperproject/prusti-dev#169]
By removing the [`prusti-filter` component](https://github.com/viperproject/prusti-dev/tree/b911f2ffa8ebb45623dc1376c6223ab21f39f81f/prusti-filter/src/validators) during the compiler upgrade, we lost many user-friendly error messages and checks of unsupported features. As those checks are no longer performed, Prusti just crashes in many cases where it used to (at least) report an error message. This issue is here to remember that we should port the old checks to the new Prusti version.

Example program:
```rust
fn test(x: u32) -> u32 {
    let f = |x| x + 1;
    f(x)
}

fn main() {}
```

Current panic, with no error messages:
```
thread 'rustc' panicked at 'not implemented', prusti-viper/src/encoder/procedure_encoder.rs:4842:19
```

Error messages generated by the old `prusti-filter` checks:
```
error: [Prusti: unsupported feature] this is unsupported, because it uses closures
 --> .../closure.rs:3:5
  |
3 |     f(x)
  |     ^

error: [Prusti: unsupported feature] this is unsupported, because it uses closures
 --> .../closure.rs:2:13
  |
2 |     let f = |x| x + 1;
  |             ^^^^^^^^^

error: [Prusti: unsupported feature] this is unsupported, because it uses closures
 --> .../closure.rs:3:5
  |
3 |     f(x)
  |     ^^^^
```

**Comment by Aurel300** [comment]
Since we don't want to bring back `prusti-filter`, I assume the solution here is to replace `unimplemented!(...)` with calls that will emit the proper Prusti error?

**Comment by fpoli** [comment]
Yes, for example returning `Err(PrustiError::unsupported(..))`. The `Encoder` should catch errors, generating stub methods/functions.

An interesting feature is that unsupported *statements* should be encoded with `assert false` Viper statements registered with an `ErrorCtxt::Unsupported(..)` context, such that we report an error message only when those statements might be reachable.

**Comment by Aurel300** [comment]
How much of this was covered by #243 and #244?

**Comment by fpoli** [comment]
> How much of this was covered by #243 and #244?

A lot, but not everything. Currently `cargo-prusti` only runs on the top-2 crates without crashing, but it used to be the top-500. The good thing is that the old filtering was extremely conservative, while the new approach is very precise and never marks code as "possibly unsupported". To close this issue, I would set as goal to run `cargo-prusti` on the top-10 crates and on Prusti itself without crashing.
